contributors/design-proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,168 @@
+
+There is an undergoing effort to move all of the volume logic from the controller-manager into plugins called Flex Volumes. In the Flex volumes world, all of the vendor specific code will be packaged in a separate binary as a plugin. After discussing with @thockin, this was decidedly the best approach to remove all cloud provider dependency for volumes out of kubernetes core.
+
+### 6. Deployment, Upgrades and Downgrades
+
+This change will introduce new binaries to the list of binaries required to run kubernetes. The change will be designed such that these binaries can be installed via `kubectl apply -f` and the appropriate instances of the binaries will be running. 
+
+##### 6.1 Upgrading kubelet and proxy
+
+The kubelet and proxy runs on every node in the kubernetes cluster. Based on your setup (systemd/other), you can follow the normal upgrade steps for it. This change does not affect the kubelet and proxy upgrade steps for your setup.
+
+##### 6.2 Upgrading plugins
+
+Plugins such as cni, flex volumes can be upgraded just as you normally upgrade them. This change does not affect the plugin upgrade steps for your setup.
+
+###### 6.3 Upgrading kubernetes core services
+
+The master node components (kube-controller-manager,kube-scheduler, kube-apiserver etc.) can be upgraded just as you normally upgrade them. This change does not affect the plugin upgrade steps for your setup.
+
+##### 6.4 Applying the cloud-controller-manager
+
+This is the only step that is different in the upgrade process. In order to complete the upgrade process, you need to apply the cloud-controller-manager deployment to the setup. A deployment descriptor file will be provided with this change. You need to apply this change using
+
+```
+kubectl apply -f cloud-controller-manager.yml
+```
+
+This will start the cloud specific controller manager in your kuberentes setup.
+
+The downgrade steps are also the same as before for all the components except the cloud-controller-manager. In case of the cloud-controller-manager, the deployment should be deleted using
+
+```
+kubectl delete -f cloud-controller-manager.yml
+```
+
+### 7. Roadmap
+
+##### 7.1 Transition plan
+
+Release 1.6: Add the first implementation of the cloud-controller-manager binary. This binary's purpose is to let users run two controller managers and address any issues that they uncover, that we might have missed. It also doubles as a reference implementation to the external cloud controller manager for the future. Since the cloud-controller-manager runs cloud specific controller loops, it is important to ensure that the kube-controller-manager does not run these loops as well. This is done by leaving the `--cloud-provider` flag unset in the kube-controller-manager.
```

Comment:
Can you flesh this out just a bit more?  Specifically - the cloud-controller-manager binary is the same control loops that were in the main controller-manager, but re-factored as described here.  Still uses the exact same cloudprovider libs.  Yes?  In other words, is it EXPECTED that everyone switches to CCM, or is it just optional and "beta" ?